Local testing

wkdewey opened the pull request:
contains all the work I've been doing on the API (see #128 which this supercedes)

wkdewey committed:
upgrade Puma

wkdewey committed:
add specific version of puma to avoid security warnings

wkdewey committed:
add facet for matching nested facet

wkdewey committed:
add filter for matching nested facet

wkdewey committed:
change split character, add missing comma

wkdewey committed:
update to later version of puma

wkdewey committed:
another round of gem updates

wkdewey committed:
parse the array for matching nested fields

wkdewey committed:
change how compound facet name is parsed

wkdewey committed:
use facet name as agg name

this is not the most semantically correct, but it doesn't break anything and orchid seems to expect it

wkdewey committed:
change query to filter

this is necessary to make this sort of query work

wkdewey committed:
fix nested filter aggregation so it doesn't cause 400 error

wkdewey committed:
check for deeper nesting of buckets

wkdewey committed:
Change separator

wkdewey committed:
Fix parsing and query for filter matching

wkdewey committed:
rewrite filtered aggregation to be either nested or not

wkdewey committed:
filtering on a single item can either be nested or not

wkdewey committed:
update config for server

wkdewey committed:
revise query to match both the facet and the filter

wkdewey committed:
use reverse nested agg for correct item count

wkdewey committed:
used doc_count from reverse nested if it exists

wkdewey committed:
change key for new elasticsearch version

wkdewey committed:
change order query to avoid deprecated '_term'

wkdewey committed:
gitignore master key

wkdewey committed:
add basic auth to elasticsearch requests

wkdewey committed:
raise number of results per facet

wkdewey committed:
use facet_limit instead of facet_num to match Orchid

wkdewey committed:
revert, will set in Orchid

wkdewey committed:
change facet_num to facet_limit

wkdewey committed:
update nested facets documentation

wkdewey committed:
add links to more detailed documentation

wkdewey committed:
clarify

wkdewey committed:
update gems

wkdewey committed:
display total search hits (above 10,000)

fixes #139

wkdewey committed:
use reverse nested on simple nested aggregations

fixes #141

wkdewey committed:
fix elasticsearch errors

wkdewey committed:
titleize bucket values because ES automatically lowercases them

wkdewey committed:
fix display of nested field facets

Need to adjust for new Elasticsearch hash formatting when finding the non-normalized facet source name

fixes #145

wkdewey committed:
handle un-normalization of values in case of array rather than string values

wkdewey committed:
change method for unnormalization of values to fix bugs

wkdewey committed:
update changelog and improve formatting

wkdewey committed:
fix YAML loading for new version of Ruby

Commit:
update Ruby and gems for new server

wkdewey committed:
handle nested field facets in case of single hash rather than array

wkdewey committed:
make sure facet key is returned if no source

wkdewey committed:
fix de-normalization for non-nested fields

fixes #148
fixes https://github.com/CDRH/african_poetics/issues/401

wkdewey committed:
sort facets properly by count when nested

fixes https://github.com/CDRH/family_letters/issues/179

wkdewey committed:
fix ES query for nested aggregations so that _source finds right item

fixes https://github.com/Willa-Cather-Archive/willa-cather-archive/issues/371

wkdewey committed:
don't try to match up non-string values

stopgap fix for https://github.com/CDRH/api/issues/151

wkdewey committed:
add a comment for future bugfixing

wkdewey committed:
handle any response error from RestClient

fixes #152

techgique commented on CHANGELOG.md:
```diff
@@ -28,65 +28,136 @@ Markdown Spec](https://github.github.com/gfm/).
+- Orchid apps that connect to the API should use `facet_limit` instead of `facet_num` in options.
+- Add nested facets as described above, if desired
+
+### Migration
```
Migration section here is duplicated.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/aac16c04f961cd4fe63b13554639c05e5e9db8b8

techgique commented on .ruby-version:
```diff
@@ -1 +1 @@
-ruby-2.6.8
+ruby-3.1.6
```
I'll bump this up to 3.1.7 to match all other recent releases.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/d6a704ebfce774df6e4db0df2022c8c1133b3ca3

techgique commented on CHANGELOG.md:
```diff
@@ -28,65 +28,136 @@ Markdown Spec](https://github.github.com/gfm/).
+  role="judge".
+- "api_version" added to all response "res" objects
+- updated documentation for new features
+- "net-smtp" gem
```
What was this gem added for? Don't think we're doing anything with email in the API. If I don't see anything using this I'll test removing it.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/a0ed47a0d7820e328e83df4cc471ceb5975f8454

techgique commented on CHANGELOG.md:
````diff
@@ -28,65 +28,136 @@ Markdown Spec](https://github.github.com/gfm/).
+  password: *****
+```
+
+to the secrets file and then close the window to save. Do not commit `config/master.key` (it should be in `gitignore`)
````
This isn't going to display well after rendered by Markdown as the code snippet in the middle will result in the text following it just printed as a regular paragraph of text. I'll reformat this.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/aac16c04f961cd4fe63b13554639c05e5e9db8b8

techgique commented on Gemfile:
```diff
@@ -26,6 +26,7 @@ gem 'puma', '~> 3.7'
 # gem 'rack-cors'
 
 gem 'bootsnap', require: false
+gem 'net-smtp'
```
Oh, maybe this was added when copying from a default new Rails app? But I don't think I've seen this gem added to any others recently. As mentioned above, I'll look to see if there is any use, otherwise I'll remove it.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/a0ed47a0d7820e328e83df4cc471ceb5975f8454

techgique commented on Gemfile.lock:
```diff
@@ -1,180 +1,200 @@
-  ruby
+  arm64-darwin-23
+  x86_64-darwin-21
+  x86_64-darwin-22
```
I think the macOS platforms being added to the platform list here caused us grief on the servers or across dev team members' workstations, so I'll probably remove these too after comparing with some other Rails app repos.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/a0ed47a0d7820e328e83df4cc471ceb5975f8454

techgique commented on app/services/search_item_req.rb:
```diff
@@ -98,15 +99,87 @@ def facets
       elsif f.include?(".")
         path = f.split(".").first
+        if f_type == "_count"
+          #make sure sort is on the acutal count of documents
```
Typo in the comment. I'll fix this.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/afaf758ea806dc302ed53fcd91d93836435b3766

techgique commented on app/services/search_item_req.rb:
```diff
@@ -98,15 +99,87 @@ def facets
+        nested = facet.include?(".")
+        if nested
+          path = facet.split(".").first
+        end
```
I think this can be merged with the `if nested` further down as there isn't any alternative use of a `path` variable.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/afaf758ea806dc302ed53fcd91d93836435b3766

techgique commented on app/services/search_item_req.rb:
```diff
@@ -161,8 +239,43 @@ def filters
+        nested = facet.include?(".")
+        if nested
+          path = facet.split(".").first
+        end
```
Same as above, this can be merged with the `if nested` below.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/afaf758ea806dc302ed53fcd91d93836435b3766

techgique commented on app/services/search_item_res.rb:
```diff
@@ -42,20 +41,44 @@ def combine_highlights
+        end
+      else
+        # if there is an array of values but no match, just return the key
+        return key
```
Unnecessary uses of `return` here, as all final lines in the control flow here are the last code for the method. I'll remove them.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/3d4842b03061d206739bc3d9fa8986afbf547f31

techgique commented on app/services/search_item_res.rb:
```diff
@@ -104,10 +126,32 @@ def reformat_facets
+      buckets = info.dig(field, field, "buckets")
+    end
+    buckets
+  end
```
This can be simplified down to just the if, elsif, and else blocks returning rather than assigning to a variable and then returning that variable.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/3d4842b03061d206739bc3d9fa8986afbf547f31

techgique commented on app/services/search_service.rb:
```diff
@@ -11,7 +11,8 @@ def initialize(url, params={}, user_req)
   def post(url_ending, json)
-    res = RestClient.post("#{@url}/#{url_ending}", json.to_json, { "content-type" => "json" } )
+    auth_hash = { "Authorization" => "Basic #{Base64::encode64("#{Rails.application.credentials.elasticsearch[:user]}:#{Rails.application.credentials.elasticsearch[:password]}")}" }
+    res = RestClient.post("#{@url}/#{url_ending}", json.to_json, auth_hash.merge({ "content-type" => "json" } ))
```
I think this can be formatted to be a little more easily read.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/b004587c12e2c66c53b164728e94b06544686d47

techgique commented on app/controllers/application_controller.rb:
```diff
@@ -3,7 +3,8 @@
   def post_search(json, error_method=method(:display_error))
-    res = RestClient.post("#{ES_URI}/_search", json.to_json, { "content-type" => "json" })
+    auth_hash = { "Authorization" => "Basic #{Base64::encode64("#{ES_USER}:#{ES_PASSWORD}")}" }
+    res = RestClient.post("#{ES_URI}/_search", json.to_json, auth_hash.merge({ "content-type" => "json" }))
```
I think this can be formatted to be a little more easily read.

techgique replied:
After looking more closely, I'm pretty sure this code isn't even being used. It's calling variables not defined anywhere (`ES_USER`, `ES_PASSWORD`, and `ES_URI`) and always calling a `raise` which would result in an exception being thrown. So I searched the code base and `post_search` isn't called from anywhere. So I think this can go.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/46d3b0db4ab229f56211afd5841db4944cc5e541

techgique commented on app/services/search_service.rb:
```diff
@@ -57,7 +58,7 @@ def search_item(id)
     if raw_res.class == RuntimeError
       on_error(raw_res, req)
-    elsif raw_res.class == RestClient::BadRequest
+    elsif raw_res.class == RestClient::ExceptionWithResponse
```
Went to look at RestClient docs to double-check this exception change (to fix an issue linked to in the commit https://github.com/CDRH/api/commit/f0b8817e5167ed960282a5f188cdacc2cab19eb4).

Re-discovered the gem hasn't been maintained for years. I made an issue about it a while ago: https://github.com/CDRH/api/issues/144 I added an additional comment related to a post I saw on the rest-client repo. But this should be a high priority for work on this repo after this release.

techgique replied:
Found this doesn't handle all exceptions during testing, so added the BadRequest exception as an `||` condition here.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/30145d1d813b61e84c50a0f881f7bf7ad9a15037

techgique commented on config/environments/development.rb:
```diff
@@ -61,4 +73,5 @@
   # CDRH CONFIGURATION
 
   config.hosts << "cdrhdev1.unl.edu"
+  config.hosts << "whitman-dev.unl.edu"
```
I'll update the dev and prod environment configs with newer standardized code.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/8e6bdca86ebb18f6e671c6aa364ea66be93439c2

techgique commented on config/initializers/new_framework_defaults_6_1.rb:
```diff
@@ -0,0 +1,67 @@
+# Be sure to restart your server when you modify this file.
+#
+# This file contains migration options to ease your Rails 6.1 upgrade.
+#
+# Once upgraded flip defaults one by one to migrate to the new default.
```
I'll enable these options and test the app.

techgique replied:
Resolved: https://github.com/CDRH/api/pull/146/commits/c42e264f4aed3d94026315fb1f505aa4c22d8d85

techgique reviewed (changes requested):
Noting what I found before I work on the code to make revisions.

techgique committed:
Bump Ruby version to 3.1.7

techgique committed:
Remove unused code in ApplicationController

techgique committed:
Reformat post code; use auth if creds present

- Make auth code more readable with less string interpolation
  and separate multiple args for methods across lines
- Check if elasticsearch credentials are present to determine whether
  to construct addition auth header and merge with json header

techgique committed:
Update Gemfile and gems

- Compare and match against a fresh `rails new --api` app
- Remove net-smtp gem from Gemfile and Changelog
- Remove extra platforms

techgique committed:
Set defaults to 5.2, enable new 6.1 defaults

Remove old new_framework_defaults initializers for 5.0 and 5.2

techgique committed:
Fix logger, listen gem broken from last update

Update of Gemfile required logger be added to boot.rb now

Listen gem version was causing Rails app boot to fail.
Hadn't run bundle update again after removing the extra platforms,
so suspect this was the problem we'd seen before with those 🤔

techgique committed:
Add standard Rails dev and prod env

Note production is modified to keep non-public production domains out of
source control

techgique committed:
Augment Elasticsearch request error handling

Add back RestClient::BadRequest exception so both display errors
from Elasticsearch in API response rather than returning
an opaque Rails error page

techgique committed:
search_item_req.rb reformatting, refactoring

- Fix typo in comment
- Reformat other comments so there is a space between # and text
- Refactor `if nested` code together and bypass extraneous variable
  when value only used once after
  - Likewise for path variable
- Fix indentation of aggregation hash

techgique committed:
Refactor search_item_res.rb

- Remove extraneous returns
- Remove extraneous use of buckets variable when direct returns work

techgique committed:
Update version, config example; review Changelog

- Version to 2.0.0
- Config example license and docs links updated. Old year removed from
  `api_update` example
- Changelog review
  - Set version for release and compare url to `v2.0.0`
  - Change explicit code references to `code` rather than "quotes"
  - Capitalize first words, remove trailing periods
  - Remove duplicate entries (`Migration` section, `api_version` added
    to `res`)
  - Drop Rails and Ruby entry as v1.0.5 for v1 already upgraded to them
  - Move text for list item including json snippets to before
    the snippets as Markdown doesn't connect the indented text after
    with prior list items
  - Add json syntax highlighting to json snippets
  - Reorder some of the major change items to the top of sections
  - Add v1.0.5 release that was forgotten with that release

techgique committed:
Merge branch 'dev' into local_testing

Resolve merge conflicts
- Keep all changes from `local_testing` for Changelog, Gemfile,
  Gemfile.lock
- Others accepted from `dev`

techgique commented:
@wkdewey, please review the changes I've made to double-check I haven't made any mistakes. We might want to test this on the dev server with the dev Rails apps utilizing it as well.

The changes I made to handle Elasticsearch both with and without auth mean we can delete the `no_security_testing` branch after this is merged and deployed. Since that's what's currently deployed, note [the differences between them before I committed changes here after my review](https://github.com/CDRH/api/compare/no_security_testing...f0b8817) and [this comparison after my changes were committed](https://github.com/CDRH/api/compare/no_security_testing...local_testing). I didn't see anything that should be kept from `no_security_testing` but let me know if you spot anything.

wkdewey commented:
The changes you made look good, but I think some testing would be necessary.

wkdewey commented:
the changes on `no_security_testing` also look good
